File: Python/Parte2.py
import pandas as pd
from wordcloud import WordCloud
import matplotlib.pyplot as plt

# Cargar los datos desde el archivo CSV
datos = pd.read_csv('web_scraping_citas.csv')

# 1. Identificar los 10 artículos con más citas
top_10_articulos = datos.sort_values(by='Citas', ascending=False).head(10)
top_10_articulos.to_csv('top_10_articulos_citados.csv', index=False)

# 2. Búsqueda de palabras clave en títulos
palabra_clave = input("Ingresa una palabra clave: ")
resultados_palabra_clave = datos[datos['Título'].str.contains(palabra_clave, case=False)]
resultados_palabra_clave.to_csv('resultados_palabra_clave.csv', index=False)

# 3. Contar todos los autores
todos_autores = datos['Autor'].str.split(' - ', expand=True).stack().value_counts().reset_index()
todos_autores.columns = ['Autor', 'Conteo']
todos_autores.to_csv('todos_autores.csv', index=False)

# 4.1. Visualización de nube de palabras
titulos = " ".join(datos['Título'])
nube_palabras = WordCloud(width=800, height=400, background_color='white').generate(titulos)
plt.figure(figsize=(10, 5))
plt.imshow(nube_palabras, interpolation='bilinear')
plt.axis('off')
plt.title('Nube de Palabras en Títulos')
plt.show()

# 4.2. Contar la cantidad de artículos por año y hacer un gráfico de barras
conteo_articulos_por_año = datos['Año'].value_counts().sort_index()
conteo_articulos_por_año.plot(kind='bar')
plt.xlabel('Año de Publicación')
plt.ylabel('Número de Artículos')
plt.title('Cantidad de Artículos por Año de Publicación')
plt.show()

File: JavaScript/Parte2.js
const fs = require('fs');
const csv = require('csv-parser');
const ObjectsToCsv = require('objects-to-csv');
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function loadAndProcessData() {
  const articles = [];
  fs.createReadStream('web_scraping_citas.csv')
    .pipe(csv())
    .on('data', (row) => {
      articles.push(row);
    })
    .on('end', () => {
      console.log('Datos cargados correctamente.');
      console.log('Elija una opción:');
      console.log('1. Identificar los 10 artículos con más citas.');
      console.log('2. Búsqueda de palabras clave en títulos.');
      console.log('3. Contar todos los autores.');
      console.log('4. Visualización gráfica.');
      rl.question('Opción: ', (option) => {
        switch (option) {
          case '1':
            top10CitedArticles(articles);
            break;
          case '2':
            searchByKeyword(articles);
            break;
          case '3':
            countAuthors(articles);
            break;
          case '4':
            visualizeData(articles);
            break;
          default:
            console.log('Opción no válida.');
        }
        rl.close();
      });
    });
}

function top10CitedArticles(articles) {
  articles.sort((a, b) => b['Citado por'] - a['Citado por']);
  const top10 = articles.slice(0, 10);
  const csv = new ObjectsToCsv(top10);
  csv.toDisk('top_10_cited_articles.csv').then(() => {
    console.log('Top 10 artículos con más citas guardados en top_10_cited_articles.csv');
  });
}

function searchByKeyword(articles) {
  rl.question('Ingrese una palabra clave: ', (keyword) => {
    const filteredArticles = articles.filter(article => article.Title.toLowerCase().includes(keyword.toLowerCase()));
    const csv = new ObjectsToCsv(filteredArticles);
    csv.toDisk('articles_with_keyword.csv').then(() => {
      console.log(`Artículos con la palabra clave "${keyword}" guardados en articles_with_keyword.csv`);
    });
  });
}

function countAuthors(articles) {
  const authorsCount = {};
  articles.forEach(article => {
    const authors = article.Author.split(', ');
    authors.forEach(author => {
      authorsCount[author] = (authorsCount[author] || 0) + 1;
    });
  });
  const authorsArray = Object.keys(authorsCount).map(author => ({ Author: author, Count: authorsCount[author] }));
  authorsArray.sort((a, b) => b.Count - a.Count);
  const csv = new ObjectsToCsv(authorsArray);
  csv.toDisk('authors_count.csv').then(() => {
    console.log('Conteo de autores guardado en authors_count.csv');
  });
}

function visualizeData(articles) {
  // Aquí puedes realizar la visualización gráfica, como generar un gráfico de nube de palabras y un gráfico de barras.
  // Puedes usar bibliotecas como D3.js, Chart.js, o cualquier otra de tu elección para crear las visualizaciones.
  console.log('Visualización gráfica no implementada en este ejemplo.');
}

loadAndProcessData();